core/metric.go

```diff
@@ -70,88 +172,100 @@ func (n Namespace) Key() string {
 func (n Namespace) AddStaticElement(value string) Namespace {
-	nse := NamespaceElement{Value: value}
-	return append(n, nse)
+	nse := StaticElement{Value_: value}
+	return append(n, &nse)
 }
 
 // AddStaticElements adds a static elements to the given Namespace.  A static
 // NamespaceElement is defined by having an empty Name field.
 func (n Namespace) AddStaticElements(values ...string) Namespace {
 	for _, value := range values {
-		n = append(n, NamespaceElement{Value: value})
+		n = append(n, &StaticElement{Value_: value})
 	}
 	return n
 }
 
+// IsDynamic returns true if there is any element of the namespace which is
+// dynamic.  If the namespace is dynamic the second return value will contain
+// an array of namespace elements (indexes) where there are dynamic namespace
+// elements. A dynamic component of the namespace are those elements that
+// contain variable data.
+func (n Namespace) IsDynamic() (bool, []int) {
+	var idx []int
+	ret := false
+	for i := range n {
+		if n[i].IsDynamic() {
+			ret = true
+			idx = append(idx, i)
+		}
+	}
+	return ret, idx
+}
+
+// Element returns the NamespaceElement at the giving index
 func (n Namespace) Element(idx int) NamespaceElement {
 	if idx >= 0 && idx < len(n) {
 		return n[idx]
 	}
-	return NamespaceElement{}
+	return &StaticElement{}
 }
 
-// NamespaceElement provides meta data related to the namespace.  This is of particular importance when
-// the namespace contains data.
-type NamespaceElement struct {
-	Value       string
-	Description string
-	Name        string
+// NamespaceJSON represents a persistable Namespace
+type NamespaceJSON struct {
+	Type string
+	Data json.RawMessage
 }
 
-// NewNamespaceElement tasks a string and returns a NamespaceElement where the
-// Value field is set to the provided string argument.
-func NewNamespaceElement(e string) NamespaceElement {
-	if e != "" {
-		return NamespaceElement{Value: e}
+// NewNamespaceJSON returns a new instance of NamespaceJSON object
+func NewNamespaceJSON(v interface{}) (*NamespaceJSON, error) {
+	b, err := json.Marshal(v)
+	if err != nil {
+		return nil, err
 	}
-	return NamespaceElement{}
+	return &NamespaceJSON{
+		reflect.Indirect(reflect.ValueOf(v)).Type().Name(), b}, nil
 }
 
-// IsDynamic returns true if the namespace element contains data.  A namespace
-// element that has a nonempty Name field is considered dynamic.
-func (n *NamespaceElement) IsDynamic() bool {
-	if n.Name != "" {
-		return true
+// MarshalJSON marshals a Namespace into JSON
+func (n Namespace) MarshalJSON() ([]byte, error) {
+	objs := []*NamespaceJSON{}
+	for _, elem := range n {
+		obj, _ := NewNamespaceJSON(elem)
+		objs = append(objs, obj)
 	}
-	return false
+	return json.Marshal(objs)
+}
+
+// UnmarshalJSON unmarshals JSON into a Namespace
+func (n Namespace) UnmarshalJSON(data []byte) error {
+	objs := []*NamespaceJSON{}
+	return json.Unmarshal(data, &objs)
+}
+
+func init() {
+	gob.Register(&StaticElement{})
+	gob.Register(&DynamicElement{})
```

Comment:
I don't know if we want to have register going on in multiple places. 

We also need to use gob.RegisterName() to avoid import path issues in the future.